### PR TITLE
[ci-visibility] Per test code coverage for `jest` 

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?

Read code coverage in `jest` if it's available (`global.__coverage__`) and reset the counters _after every test_. This results in a code coverage that is exclusive to the test. This works because tests (`it` clauses) are run sequentially within a test file. 

After code coverage is formatted, it is exported through `exportCoverage`, a new function in `AgentlessCiVisibilityExporter`, which ends up using `CIVisibilityCoverageEncoder` encoder, which encodes the data in the expected format. 

Additionally, a new dependency `istanbul-lib-coverage` is added for parsing the code coverage: 
https://www.npmjs.com/package/istanbul-lib-coverage
https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-coverage 

### Motivation
Per test code coverage is another requirement for intelligent test runner. 

### Plugin Checklist

- [x] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/datadog-plugin-jest/src/index.js [hunk 1]
```diff
@@ -41,6 +41,14 @@ class JestPlugin extends Plugin {
     this.testEnvironmentMetadata = getTestEnvironmentMetadata('jest', this.config)
     this.codeOwnersEntries = getCodeOwnersFileEntries()
 
+    this.addSub('ci:jest:test:code-coverage', (coverageFiles) => {
+      if (!this.config.isAgentlessEnabled || !this.config.isITREnabled) {
+        return
+      }
+      const testSpan = storage.getStore().span
+      this.tracer._exporter.exportCoverage({ testSpan, coverageFiles })
```
Comment (juan-fernandez) [inline packages/datadog-plugin-jest/src/index.js:49]: 💡 we could expose this function in the tracer directly

File: packages/dd-trace/src/encode/agentless-ci-visibility.js [hunk 2]
```diff
@@ -29,7 +29,6 @@ function formatSpan (span) {
 class AgentlessCiVisibilityEncoder extends AgentEncoder {
   constructor ({ runtimeId, service, env }) {
     super(...arguments)
-    this._events = []
```
Comment (juan-fernandez) [inline packages/dd-trace/src/encode/agentless-ci-visibility.js:32]: wasn't being used

File: packages/dd-trace/src/exporters/common/request.js [hunk 3]
```diff
@@ -74,7 +83,11 @@ function request (data, options, keepAlive, callback) {
       onError(err)
     })
 
-    dataArray.forEach(buffer => req.write(buffer))
+    if (isFormData) {
+      data.pipe(req)
+    } else {
+      dataArray.forEach(buffer => req.write(buffer))
+    }
```
Comment (juan-fernandez) [inline packages/dd-trace/src/exporters/common/request.js:90]: the changes here is so that we can use `request` with an instance of `FormData`

File: packages/datadog-plugin-jest/src/index.js [hunk 4]
```diff
@@ -41,6 +41,14 @@ class JestPlugin extends Plugin {
     this.testEnvironmentMetadata = getTestEnvironmentMetadata('jest', this.config)
     this.codeOwnersEntries = getCodeOwnersFileEntries()
 
+    this.addSub('ci:jest:test:code-coverage', (coverageFiles) => {
+      if (!this.config.isAgentlessEnabled || !this.config.isITREnabled) {
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:45]: ITR?

File: packages/dd-trace/src/encode/0.4.js [hunk 5]
```diff
@@ -126,6 +126,19 @@ class AgentEncoder {
     buffer[offset + 4] = length
   }
 
+  _encodeMapPrefix (bytes, keysLength) {
```
Comment (rochdev) [inline packages/dd-trace/src/encode/0.4.js:129]: Should this be used in `encodeMap` as well?

File: packages/dd-trace/src/exporters/common/request.js [hunk 6]
```diff
@@ -15,20 +16,28 @@ const httpAgent = new http.Agent({ keepAlive, maxTotalSockets })
...
+}
+
 let activeRequests = 0
 
 function request (data, options, keepAlive, callback) {
   if (!options.headers) {
     options.headers = {}
   }
 
+  const isFormData = isForm(data)
```
Comment (rochdev) [inline packages/dd-trace/src/exporters/common/request.js:30]: Why not just inline the function? It checks if it's readable, not if it's FormData.

File: packages/dd-trace/src/exporters/common/request.js (hunk 3 above)
Comment (rochdev) [reply packages/dd-trace/src/exporters/common/request.js:90]: Let's do something like `isStream` of `isReadable` instead since it doesn't necessarily need to be form data.

File: packages/dd-trace/src/exporters/common/request.js [hunk 7]
```diff
@@ -74,10 +83,16 @@ function request (data, options, keepAlive, callback) {
...
-    dataArray.forEach(buffer => req.write(buffer))
+    if (isFormData) {
+      data.pipe(req)
+    } else {
+      dataArray.forEach(buffer => req.write(buffer))
+    }
 
     req.setTimeout(timeout, req.abort)
-    req.end()
+    if (!isFormData) {
```
Comment (rochdev) [inline packages/dd-trace/src/exporters/common/request.js:93]: If the previous statement is moved up then in could be in the same condition.

File: packages/dd-trace/src/plugin_manager.js [hunk 8]
```diff
@@ -134,9 +135,12 @@ module.exports = class PluginManager {
     }
 
     // TODO: update so that it's available for every CI Visibility's plugin
-    if (name === 'mocha') {
+    if (name === 'mocha' || name === 'jest') {
```
Comment (rochdev) [inline packages/dd-trace/src/plugin_manager.js:138]: Why is this needed actually? Could the option just be passed to everything like the other options? Then any plugin that supports agentless can see the config.

File: packages/dd-trace/src/plugin_manager.js [hunk 9]
```diff
@@ -134,9 +135,12 @@ module.exports = class PluginManager {
...
       sharedConfig.isAgentlessEnabled = experimental && experimental.exporter === 'datadog'
     }
+    if (name === 'jest') {
```
Comment (rochdev) [inline packages/dd-trace/src/plugin_manager.js:141]: Similar to the above, why isolate the config here? Also, what is this config in the first place?

File: packages/dd-trace/src/ci-visibility/exporters/agentless/writer.js [hunk 10]
```diff
@@ -24,28 +52,48 @@ class Writer extends BaseWriter {
...
+      '/api/v2/citestcycle',
+      { 'Content-Type': 'application/msgpack' }
+    )
+    this._sendPayloadBase(data, options, done)
   }
 
-  options.protocol = url.protocol
-  options.hostname = url.hostname
-  options.port = url.port
+  _sendCoverage (form, done) {
```
Comment (rochdev) [inline packages/dd-trace/src/ci-visibility/exporters/agentless/writer.js:65]: My original issue with having separate writers was not having multiple classes, but having 2 completely distinct ways to write to the intake. Since this has been implemented exactly this way, multiple writers would make more sense until we can come up with a normalized writer.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-plugin-jest/src/index.js (hunk 4 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:45]: intelligent test runner

File: packages/dd-trace/src/plugin_manager.js (hunk 8 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/plugin_manager.js:138]: yeah it could be. I thought it'd be best not to affect other plugins, but this is innocuous

File: packages/datadog-plugin-jest/src/index.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-jest/src/index.js:45]: Could we use some terminology that is not an unknown acronym to make this easier to use?

Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:45]: yeah that makes sense

**Comment by juan-fernandez** [comment]
any other concern @rochdev ? I think I applied all the feedback

File: packages/datadog-plugin-jest/src/index.js (hunk 4 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js]: done

File: packages/dd-trace/src/ci-visibility/exporters/agentless/index.js [hunk 11]
```diff
@@ -24,8 +57,17 @@ class AgentlessCiVisibilityExporter {
...
+  setUrl (url, coverageUrl = url) {
+    try {
+      url = new URL(url)
+      coverageUrl = new URL(url)
+      this._url = url
+      this._coverageUrl = coverageUrl
+      this._writer.setUrl(url)
+      this._coverageWriter.setUrl(coverageUrl)
+    } catch (e) {
+      log.warn(e.stack)
```
Comment (rochdev) [inline packages/dd-trace/src/ci-visibility/exporters/agentless/index.js:69]: Let's use an error log since that handles error properties already.

File: packages/dd-trace/src/ci-visibility/exporters/agentless/index.js [hunk 12]
```diff
@@ -2,18 +2,51 @@
...
       this._scheduler = new Scheduler(() => this._writer.flush(), flushInterval)
+
+      if (this._isIntelligentTestRunnerEnabled) {
+        this._coverageScheduler = new Scheduler(() => this._coverageWriter.flush())
+      }
     }
+
     this._scheduler && this._scheduler.start()
+
+    // Reduce likelihood of requests overlapping
```
Comment (rochdev) [inline packages/dd-trace/src/ci-visibility/exporters/agentless/index.js:31]: Why is this needed?

File: packages/dd-trace/src/ci-visibility/exporters/agentless/index.js (hunk 11 above)
Comment (rochdev) [reply packages/dd-trace/src/ci-visibility/exporters/agentless/index.js:69]:
```suggestion
      log.error(e)
```

File: packages/dd-trace/src/exporters/common/request.js [hunk 13]
```diff
@@ -22,13 +23,17 @@ function request (data, options, keepAlive, callback) {
...
+
   // The timeout should be kept low to avoid excessive queueing.
   const timeout = options.timeout || 2000
   const isSecure = options.protocol === 'https:'
   const client = isSecure ? https : http
   const dataArray = [].concat(data)
 
-  options.headers['Content-Length'] = byteLength(dataArray)
+  if (!isReadable) {
+    options.headers['Content-Length'] = byteLength(dataArray)
```
Comment (rochdev) [inline packages/dd-trace/src/exporters/common/request.js:35]: Any way to know the content length even with a stream? It's unclear if this was added because of an issue in the past or just as best practice.

File: packages/dd-trace/src/ci-visibility/exporters/agentless/index.js (hunk 12 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/ci-visibility/exporters/agentless/index.js:31]: I figured it'd be better if we don't fire two requests at the same time. I can remove it if you think it doesn't add much. In any case, after we merge https://github.com/DataDog/dd-trace-js/pull/2296 this won't look like this

File: packages/dd-trace/src/exporters/common/request.js (hunk 13 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/exporters/common/request.js:35]: not that I know

**Review by rochdev — APPROVED** [review]

File: packages/datadog-plugin-jest/test/circus.spec.js [hunk 14]
```diff
@@ -236,6 +251,52 @@ describe('Plugin', function () {
...
+              .to.include('packages/datadog-plugin-jest/test/sum-coverage-test.js')
+            expect(contentDisposition).to.contain(
+              'Content-Disposition: form-data; name="coverage1"; filename="coverage1.msgpack"'
+            )
+            expect(eventContentDisposition).to.contain(
+              'Content-Disposition: form-data; name="event"; filename="event.json"'
+            )
+            expect(eventPayload).to.equal('{}')
+            done()
+          }, 2000)
```
Comment (Qard) [inline packages/datadog-plugin-jest/test/circus.spec.js:296]: Rather than a timeout, couldn't you just put all these checks inside the nock reply handler?

File: packages/dd-trace/src/ci-visibility/exporters/agentless/index.js [hunk 15]
```diff
@@ -28,6 +59,19 @@ class AgentlessCiVisibilityExporter {
       }, flushInterval).unref()
     }
   }
+
+  setUrl (url, coverageUrl = url) {
+    try {
+      url = new URL(url)
+      coverageUrl = new URL(url)
```
Comment (Qard) [inline packages/dd-trace/src/ci-visibility/exporters/agentless/index.js:66]:
```suggestion
      coverageUrl = new URL(coverageUrl)
```

File: packages/dd-trace/src/ci-visibility/exporters/agentless/writer.js [hunk 16]
```diff
@@ -14,7 +20,22 @@ class Writer extends BaseWriter {
   }
 
   _sendPayload (data, _, done) {
-    makeRequest(data, this._url, (err, res) => {
+    const options = {
+      path: '/api/v2/citestcycle',
+      method: 'POST',
+      headers: {
+        'dd-api-key': process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
```
Comment (Qard) [inline packages/dd-trace/src/ci-visibility/exporters/agentless/writer.js:27]: What happens if neither env var is defined?

File: packages/dd-trace/test/ci-visibility/exporters/agentless/exporter.spec.js [hunk 17]
```diff
@@ -58,4 +66,40 @@ describe('CI Visibility Exporter', () => {
...
+      }
+      const payload = { testSpan, coverageFiles: ['file.js'] }
+
+      exporter.exportCoverage(payload)
+
+      setTimeout(() => {
+        expect(coverageWriter.flush).to.have.been.called
+        done()
+      }, flushInterval)
+      expect(coverageWriter.append).to.have.been.called
```
Comment (Qard) [inline packages/dd-trace/test/ci-visibility/exporters/agentless/exporter.spec.js:102]: Could use a check to make sure it's not called immediately.

```suggestion
      expect(coverageWriter.append).to.have.been.called
      expect(coverageWriter.flush).to.not.have.been.called
```

File: packages/dd-trace/src/ci-visibility/exporters/agentless/index.js (hunk 15 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/ci-visibility/exporters/agentless/index.js:66]: good catch

File: packages/dd-trace/src/ci-visibility/exporters/agentless/writer.js (hunk 16 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/ci-visibility/exporters/agentless/writer.js:27]: agentless is not activated: https://github.com/DataDog/dd-trace-js/blob/master/ci/init.js#L18, so we wouldn't reach this point

**Review by Qard — APPROVED** [review]
